kythe/rust/indexer/src/bin/proxy/main.rs

```diff
@@ -0,0 +1,119 @@
+
+        // Index the CompilationUnit and let the proxy know we are done
+        match indexer.index_cu(&unit, &temp_path, &mut file_provider) {
+            Ok(_) => {
+                println!(r#"{{"req":"done", "args"{{"ok":"true","msg":""}}}}"#);
+            },
+            Err(err) => {
+                println!(r#"{{"req":"done", "args"{{"ok":"false","msg":"{:?}"}}}}"#, err);
+            },
+        };
+
+        // Grab the response, but we don't care what it is so just throw it away
+        let mut response_string = String::new();
+        io::stdin()
+            .read_line(&mut response_string)
+            .context("Failed to read response from proxy")?;
+    }
+}
+
+/// Takes analysis files present in a CompilationUnit, requests the files
+/// from the proxy, and writes them to `temp_path`
+pub fn write_analysis_to_directory(
+    c_unit: &CompilationUnit,
+    temp_path: &Path,
+    provider: &mut dyn FileProvider,
+) -> Result<()> {
+    for required_input in c_unit.get_required_input() {
+        let input_path = required_input.get_info().get_path();
+        let input_path_buf = PathBuf::new().join(input_path);
+
+        // save_analysis files are JSON files
+        if let Some(os_str) = input_path_buf.extension() {
+            if let Some("json") = os_str.to_str() {
+                let digest = required_input.get_info().get_digest();
+                let file_contents = provider.contents(&input_path, digest).with_context(|| {
+                    format!(
+                        "Failed to get contents of file \"{}\" with digest \"{}\"",
+                        input_path, digest
+                    )
+                })?;
+
+                let output_path = temp_path.join(input_path_buf.file_name().unwrap());
+                let mut output_file = File::create(&output_path).context("Failed to create file")?;
+                output_file.write_all(&file_contents).with_context(|| {
+                    format!(
+                        "Failed to copy contents of \"{}\" with digest \"{}\" to \"{}\"",
+                        input_path,
+                        digest,
+                        output_path.display()
+                    )
+                })?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn request_compilation_unit() -> Result<CompilationUnit> {
+    println!(r#"{{"req":"analysis_wire"}}"#);
+    io::stdout().flush().context("Failed to flush stdout")?;
+
+    // Read the response from the proxy
+    let mut response_string = String::new();
+    io::stdin()
+        .read_line(&mut response_string)
+        .context("Failed to read response from proxy")?;
+
+    // Convert to json and extract information
+    let response: Value = serde_json::from_str(&response_string)
+        .context("Failed to parse response as JSON")?;
+    if response["rsp"].as_str().unwrap() == "ok" {
+        let args = response["args"].clone();
+        let unit_base64 = args["unit"].as_str().unwrap();
```

Comment:
> Since this is a fatal error, shouldn't the indexer simply die, which will be detected by the proxy?

Yes, that's true. I'm just confirming that if there's no "unit" field, then the indexer is right to decide "the proxy is misbehaving so I should panic". That does seem to make sense.